ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/HARouting.java

```diff
@@ -0,0 +1,318 @@
+   * of host nodes.
+   * @param statement the statement from which this request came
+   * @param locations the list of partition locations to parse
+   * @param round which round this is
+   * @return A map of node to list of partition locations
+   */
+  private Map<KsqlNode, List<KsqlPartitionLocation>> groupByHost(
+      final ConfiguredStatement<Query> statement,
+      final List<KsqlPartitionLocation> locations,
+      final int round) {
+    final Map<KsqlNode, List<KsqlPartitionLocation>> groupedByHost = new LinkedHashMap<>();
+    for (KsqlPartitionLocation location : locations) {
+      // If one of the partitions required is out of nodes, then we cannot continue.
+      if (round >= location.getNodes().size()) {
+        throw new MaterializationException(String.format(
+            "Unable to execute pull query: %s. Exhausted standby hosts to try.",
+            statement.getStatementText()));
+      }
+      final KsqlNode nextHost = location.getNodes().get(round);
+      groupedByHost.computeIfAbsent(nextHost, h -> new ArrayList<>()).add(location);
+    }
+    return groupedByHost;
+  }
+
+  @VisibleForTesting
+  interface RouteQuery {
+    PullQueryResult routeQuery(
+        KsqlNode node,
+        List<KsqlPartitionLocation> locations,
+        ConfiguredStatement<Query> statement,
+        ServiceContext serviceContext,
+        RoutingOptions routingOptions,
+        Optional<PullQueryExecutorMetrics> pullQueryMetrics
+    );
+  }
+
+  private PullQueryResult routeQuery(
+      final KsqlNode node,
+      final List<KsqlPartitionLocation> locations,
+      final ConfiguredStatement<Query> statement,
+      final ServiceContext serviceContext,
+      final RoutingOptions routingOptions,
+      final Optional<PullQueryExecutorMetrics> pullQueryMetrics
+  ) {
+    List<List<?>> rows = null;
+    if (node.isLocal()) {
+      LOG.debug("Query {} executed locally at host {} at timestamp {}.",
+                statement.getStatementText(), node.location(), System.currentTimeMillis());
+      pullQueryMetrics
+          .ifPresent(queryExecutorMetrics -> queryExecutorMetrics.recordLocalRequests(1));
+      rows = pullPhysicalPlan.execute(locations, pullQueryMetrics);
+
+    } else {
+      LOG.debug("Query {} routed to host {} at timestamp {}.",
+                statement.getStatementText(), node.location(), System.currentTimeMillis());
+      pullQueryMetrics
+          .ifPresent(queryExecutorMetrics -> queryExecutorMetrics.recordRemoteRequests(1));
+      rows = forwardTo(node, locations, statement, serviceContext);
+    }
+    final Optional<List<KsqlNode>> debugNodes = Optional.ofNullable(
+        routingOptions.isDebugRequest()
+            ? Collections.nCopies(rows.size(), node) : null);
+    return new PullQueryResult(
+        rows,
+        debugNodes,
+        outputSchema,
+        queryId);
+  }
+
+  private static List<List<?>> forwardTo(
```

Comment:
Currently, we do multiple requests in parallel using the executorService.  We should maybe make this logic use async rather than a threadpool and futures, but it accomplishes the same thing at the moment.

Reply:
Yeah I agree the current code achieves the same thing, but using a threadpool of 100 threads by default could be costly than using a single thread (or at most, we only need one thread per destination socket). Anyways, as I mentioned in the meta comment I feel this is not necessarily the scope of the PR and we can just refactor it later if people agree to the approach.